### PR TITLE
Data loading

### PR DESCRIPTION
**Pull request by ksew1**
Closes https://github.com/software-mansion/cairo-coverage/issues/9

### CODE REVIEW

File: crates/cairo-coverage/src/input/mappings/statement_map.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+fn find_statement_origin(
+    code_locations: &[CodeLocation],
+    function_names: &[FunctionName],
+) -> Option<StatementOrigin> {
+    code_locations
+        .iter()
+        .zip(function_names)
+        // TODO: We should probably filter by path to user project not by path to cache
```
Comment (ksew1) [inline crates/cairo-coverage/src/input/mappings/statement_map.rs:60]: Will be done in separate PR

Comment (MaksymilianDemitraszek) [reply crates/cairo-coverage/src/input/mappings/statement_map.rs:60]: Is there an issue?

File: crates/cairo-coverage/src/input/data.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+use cairo_lang_sierra::program::{ProgramArtifact, VersionedProgram};
+use camino::Utf8PathBuf;
+use std::borrow::Borrow;
+use trace_data::CallTrace;
+
+#[allow(dead_code)] // Temporary
+pub struct InputData {
+    pub sierra_ids: SierraIds,
+    pub statement_map: StatementMap,
+}
```
Comment (MaksymilianDemitraszek) [inline crates/cairo-coverage/src/input/data.rs:13]: No idea what this structure represents.

File: crates/cairo-coverage/src/input/mappings/statement_map.rs (hunk 1 above)
Comment (ksew1) [reply crates/cairo-coverage/src/input/mappings/statement_map.rs:60]: https://github.com/software-mansion/cairo-coverage/issues/26

File: crates/cairo-coverage/src/input/data_loader/types.rs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.start.line..=self.end.line
+    }
+}
+
+#[derive(Deserialize, Clone, Eq, PartialEq)]
+pub struct Position {
+    pub line: LineNumber,
+}
```
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/cairo-coverage/src/input/data_loader/types.rs:27]:
```json
              {
                "end": {
                  "col": 4,
                  "line": 3
                },
                "start": {
                  "col": 4,
                  "line": 3
                }
              }
```

The structure of json is nested so you need a nested struct

File: crates/cairo-coverage/src/input/data_loader/types.rs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+#[derive(Deserialize, Clone, Eq, PartialEq)]
+pub struct LineRange {
+    pub start: Position,
+    pub end: Position,
+}
+
+impl IntoIterator for &LineRange {
+    type Item = LineNumber;
+    type IntoIter = RangeInclusive<LineNumber>;
```
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/cairo-coverage/src/input/data_loader/types.rs:17]: Yep, 

```json
              {
                "end": {
                  "col": 4,
                  "line": 3
                },
                "start": {
                  "col": 4,
                  "line": 3
                }
              }
```

This is needs to be inclusive otherwise it doesn't make sense

File: crates/cairo-coverage/src/types.rs
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/cairo-coverage/src/types.rs:1]: It will be latter used in other parts of the code and is not strictly tied to data loader

File: crates/cairo-coverage/src/input/mappings/sierra_ids.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+use cairo_lang_sierra_to_casm::metadata::{calc_metadata, MetadataComputationConfig};
+use derived_deref::Deref;
+use std::collections::HashSet;
+use trace_data::{CallTrace, CasmLevelInfo};
+
+#[derive(Deref)]
+pub struct SierraIds(HashSet<StatementIdx>);
+
+impl SierraIds {
+    pub fn new(program: &Program, call_trace: &CallTrace) -> Result<Self> {
```
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/cairo-coverage/src/input/mappings/sierra_ids.rs:13]: It maps casm ids to sierra ids which can be used latter to gets keys of statement map

File: crates/cairo-coverage/src/input/data_loader/deserialize.rs [hunk 6]
```diff
@@ -0,0 +1,26 @@
+use anyhow::{Context, Result};
+use cairo_lang_sierra::debug_info::Annotations;
+use camino::Utf8PathBuf;
+use serde::de::DeserializeOwned;
+use std::fs;
+
+pub fn from_file<T: DeserializeOwned>(file_path: &Utf8PathBuf) -> Result<T> {
```
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/cairo-coverage/src/input/data_loader/deserialize.rs:7]: It was meant to be used  with module like this `deserialize::from_file()` to omit clippy warning `clippy::module_name_repetitions`

File: crates/cairo-coverage/src/input/data_loader/deserialize.rs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    fs::read_to_string(file_path)
+        .context(format!("Failed to read file at path: {file_path}"))
+        .and_then(|content| {
+            serde_json::from_str(&content).context(format!(
+                "Failed to deserialize JSON content from file at path: {file_path}"
+            ))
+        })
+}
+
+pub fn by_namespace<T: DeserializeOwned>(annotations: &Annotations, namespace: &str) -> Result<T> {
```
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/cairo-coverage/src/input/data_loader/deserialize.rs:17]: It was meant to be used  with module like this `deserialize::by_namespace()` to omit clippy warning `clippy::module_name_repetitions`

File: crates/cairo-coverage/src/input/mappings/sierra_ids.rs [hunk 8]
```diff
@@ -0,0 +1,60 @@
+use anyhow::{Context, Result};
+use cairo_lang_sierra::program::{Program, StatementIdx};
+use cairo_lang_sierra_to_casm::compiler::{CairoProgram, SierraToCasmConfig};
+use cairo_lang_sierra_to_casm::metadata::{calc_metadata, MetadataComputationConfig};
+use derived_deref::Deref;
+use std::collections::HashSet;
+use trace_data::{CallTrace, CasmLevelInfo};
+
+#[derive(Deref)]
+pub struct SierraIds(HashSet<StatementIdx>);
```
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/cairo-coverage/src/input/mappings/sierra_ids.rs:10]: executed  sierra ids. Maybe lets name this `ExecutedSierraIds` ?

File: crates/cairo-coverage/src/input/mappings/sierra_ids.rs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+impl SierraIds {
+    pub fn new(program: &Program, call_trace: &CallTrace) -> Result<Self> {
+        let casm = compile_to_casm(program)?;
```
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/cairo-coverage/src/input/mappings/sierra_ids.rs:14]: Will be moved to input data

File: crates/cairo-coverage/src/types.rs [hunk 10]
```diff
@@ -0,0 +1,3 @@
+pub type FileLocation = String;
+pub type FunctionName = String;
+pub type LineNumber = usize;
```
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/cairo-coverage/src/types.rs:3]: Could you explain what the advantages of using them are? I don't see it.

[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply crates/cairo-coverage/src/types.rs:3]: Here, you can use FileLocation in place when `FunctionName` is required, otherwise type system guarantees it is correct.

File: crates/cairo-coverage/src/types.rs
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply crates/cairo-coverage/src/types.rs:1]: After browsing the subsequent PRs I agree, it is ok to keep it separate.

File: crates/cairo-coverage/src/input/data_loader/types.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply crates/cairo-coverage/src/input/data_loader/types.rs:27]: Ah, ok true, it is for parsing purposes.

File: crates/cairo-coverage/src/input/mappings/sierra_ids.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply crates/cairo-coverage/src/input/mappings/sierra_ids.rs:13]: Yea, I think this is a struct naming issue. The one you proposed makes it clear.

File: crates/cairo-coverage/src/types.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/cairo-coverage/src/types.rs:3]: I see your point. However, you'll either have to consistently use the `.0` accessor or both derive and implement multiple methods and traits, which can be cumbersome as these are really basic types.

 In the example, putting `FileLocation` in place of `FunctionName` would not change the behavior of the program, as they both are strings.

I still lean towards type aliases, but I don't have a really strong opinion on it. If you really think it is a better idea, we can go with newtypes.

File: crates/cairo-coverage/src/input/mappings/statement_map.rs [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+use cairo_lang_sierra::debug_info::Annotations;
+use cairo_lang_sierra::program::StatementIdx;
+use derived_deref::Deref;
+use std::collections::HashMap;
+
+const PROFILER_NAMESPACE: &str = "github.com/software-mansion/cairo-profiler";
+const COVERAGE_NAMESPACE: &str = "github.com/software-mansion/cairo-coverage";
+
+#[derive(Deref)]
+pub struct StatementMap(HashMap<StatementIdx, StatementOrigin>);
```
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/cairo-coverage/src/input/mappings/statement_map.rs:16]: Well, the real reason we created the newtype here is that I wanted to use `From` to make it consistent with the architecture here, which in transition before there different states: input -> coverage_data -> output (currently lcov, but this will be a place for other formats if we wish for it). So, tl;dr, to use `From`.

Maybe we can work with naming to still use this as the newtype? Like suggested `StatementsMap` or `StatementsById`?

File: crates/cairo-coverage/src/input/data_loader/types.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/cairo-coverage/src/input/data_loader/types.rs:17]: https://github.com/software-mansion/cairo-coverage/issues/27

File: crates/cairo-coverage/src/types.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply crates/cairo-coverage/src/types.rs:3]: I don't think this matter that much, OK with merging as it is.

File: crates/cairo-coverage/src/input/mappings/statement_map.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply crates/cairo-coverage/src/input/mappings/statement_map.rs:16]: Use functions instead traits imo, From trait does not bring much value here as it is a core bussines logic, not utility which will be reuse across the codebase.

File: crates/cairo-coverage/src/input/statements_map.rs [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+};
+use crate::types::{FileLocation, FunctionName};
+use anyhow::{Context, Result};
+use cairo_lang_sierra::debug_info::Annotations;
+use cairo_lang_sierra::program::StatementIdx;
+use serde::de::DeserializeOwned;
+use std::collections::HashMap;
+
+const PROFILER_NAMESPACE: &str = "github.com/software-mansion/cairo-profiler";
+const COVERAGE_NAMESPACE: &str = "github.com/software-mansion/cairo-coverage";
```
Comment (MaksymilianDemitraszek) [inline crates/cairo-coverage/src/input/statements_map.rs:12]: Would be better to implement those as methods of a trait `Namespace` with method `key`(names tbd) and implement those for `CoverageAnnotations` and `ProfilerAnnotations` as those are tied. It is ok to leave as it is, just knowledge sharing.

**Review by MaksymilianDemitraszek — APPROVED** [review]

File: crates/cairo-coverage/src/input/data.rs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+            .context("Failed calculating Sierra variables")?,
+        SierraToCasmConfig {
+            gas_usage_check: false,
+            max_bytecode_size: usize::MAX,
+        },
+    )
+    .context("Failed to compile sierra to casm")
+}
+
+fn read_from_file<T: DeserializeOwned>(file_path: &Utf8PathBuf) -> Result<T> {
```
Comment (MaksymilianDemitraszek) [inline crates/cairo-coverage/src/input/data.rs:66]:
```suggestion
fn read_and_serialize<T: DeserializeOwned>(file_path: &Utf8PathBuf) -> Result<T> {
```

File: crates/cairo-coverage/src/input/statements_map.rs [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+pub type StatementsMap = HashMap<StatementIdx, StatementOrigin>;
```
Comment (MaksymilianDemitraszek) [inline crates/cairo-coverage/src/input/statements_map.rs:14]:
```suggestion
pub type SierraToCairoMap = HashMap<StatementIdx, StatementOrigin>;
```
if you want to create an alias. It is better to use a newtype.

File: crates/cairo-coverage/src/input/executed_sierra_ids.rs [hunk 15]
```diff
@@ -0,0 +1,44 @@
+use anyhow::{Context, Result};
+use cairo_lang_sierra::program::StatementIdx;
+use cairo_lang_sierra_to_casm::compiler::CairoProgram;
+use derived_deref::Deref;
+use std::collections::HashSet;
+use trace_data::{CallTrace, CasmLevelInfo};
+
+#[derive(Deref)]
+pub struct ExecutedSierraIds(HashSet<StatementIdx>);
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply crates/cairo-coverage/src/input/executed_sierra_ids.rs:9]: Actually, don't you lose some information here? What if a statement is executed multiple times?
